contrib/lib/src/compression/fairing.rs

```diff
@@ -0,0 +1,150 @@
+/// is preferred over gzip.
+///
+/// In the brotli compression mode (using the
+/// [rust-brotli](https://github.com/dropbox/rust-brotli) crate), quality is set
+/// to 2 in order to achieve fast compression with a compression ratio similar
+/// to gzip. When appropriate, brotli's text and font compression modes are
+/// used.
+///
+/// In the gzip compression mode (using the
+/// [flate2](https://github.com/alexcrichton/flate2-rs) crate), quality is set
+/// The Compression type implements brotli and gzip compression for responses in
+/// to the default (9) in order to have good compression ratio.
+///
+/// This fairing does not compress responses that already have a
+/// `Content-Encoding` header.
+///
+/// This fairing ignores the responses with a `Content-Type` matching any of
+/// the following default types:
+///
+/// - application/gzip
+/// - application/brotli
+/// - application/zip
+/// - image/*
+/// - application/wasm
+/// - application/binary
+///
+/// The excluded types can be changed changing the `compress.exclude` Rocket
+/// configuration property.
+///
+/// # Usage
+///
+/// To use, add the `brotli_compression` feature, the `gzip_compression`
+/// feature, or the `compression` feature (to enable both algorithms) to the
+/// `rocket_contrib` dependencies section of your `Cargo.toml`:
+///
+/// ```toml,ignore
+/// [dependencies.rocket_contrib]
+/// version = "*"
+/// default-features = false
+/// features = ["compression"]
+/// ```
+///
+/// Then, ensure that the compression [fairing](/rocket/fairing/) is attached to
+/// your Rocket application:
+///
+/// ```rust
+/// extern crate rocket;
+/// extern crate rocket_contrib;
+///
+/// use rocket_contrib::compression::Compression;
+///
+/// fn main() {
+///     rocket::ignite()
+///         // ...
+///         .attach(Compression::fairing())
+///         // ...
+///     # ;
+/// }
+/// ```
+pub struct Compression(());
+
+impl Compression {
+    /// Returns a fairing that compresses outgoing requests.
+    ///
+    /// ## Example
+    /// To attach this fairing, simply call `attach` on the application's
+    /// `Rocket` instance with `Compression::fairing()`:
+    ///
+    /// ```rust
+    /// extern crate rocket;
+    /// extern crate rocket_contrib;
+    ///
+    /// use rocket_contrib::compression::Compression;
+    ///
+    /// fn main() {
+    ///     rocket::ignite()
+    ///         // ...
+    ///         .attach(Compression::fairing())
+    ///         // ...
+    ///     # ;
+    /// }
+    /// ```
+    pub fn fairing() -> Compression {
+        Compression { 0: () }
+    }
+}
+
+impl Fairing for Compression {
+    fn info(&self) -> Info {
+        Info {
+            name: "Response compression",
+            kind: Kind::Attach | Kind::Response,
+        }
+    }
+
+    fn on_attach(&self, rocket: Rocket) -> Result<Rocket, Rocket> {
+        let mut ctxt = Context::new();
+        match rocket.config().get_slice("compress.exclude") {
```

Comment:
Does compress.exclude work here? I would expect that you would have to do something like `get_table("compress").and_then(|t| t.get_slice("exclude"))`. I don't see any tests for the configuration syntax, which could probably go in one of the existing top-level examples (maybe `static_files`?)

Reply:
I have only tested it putting "compress.exclude" in tests, so I have no idea. @SergioBenitez, how do you think that it should be done?

Reply:
In TOML, these are all equivalent (unless I made a mistake somewhere):
```toml
[global]
compress.exclude = ["text/html"]

[global]
compress = { exclude = ["text/html"] }

[global.compress]
exclude = ["text/html"]
```

For `get_slice("compress.exclude")` to work, this syntax is required:
```toml
[global]
"compress.exclude" = ["text/html"]
```
which I think is surprising to users.